libeos-updater-util/flatpak.c

```diff
@@ -0,0 +1,1411 @@
+                                            GError **error)
+{
+  g_autoptr(FlatpakInstallation) installation = eos_updater_get_flatpak_installation (NULL, error);
+  g_autoptr(GError) local_error = NULL;
+  g_autofree gchar *value = NULL;
+
+  g_assert (out_strv != NULL);
+
+  if (!installation)
+    return FALSE;
+
+  /* TODO: Right now this returns a only the testing override or NULL,
+   * but we might want to do something a little more clever based on what is
+   * supported by Flatpak in future */
+  *out_strv = eos_updater_override_locales_list ();
+
+  return TRUE;
+}
+
+static gboolean
+locale_skip_filter_applies (JsonNode  *filter_value,
+                            gboolean  *is_filtered,
+                            GError   **error)
+{
+  JsonArray *skip_locales_array = json_node_get_array (filter_value);
+  g_autoptr(GList) locale_nodes = NULL;
+  GList *iter = NULL;
+  g_auto(GStrv) supported_languages = NULL;
+
+  g_return_val_if_fail (is_filtered != NULL, FALSE);
+
+  if (!get_locales_list_from_flatpak_installation (&supported_languages,
+                                                   error))
+    return FALSE;
+
+  if (skip_locales_array == NULL)
+    {
+      g_autofree gchar *node_str = json_node_to_string (filter_value);
+      g_set_error (error,
+                   EOS_UPDATER_ERROR,
+                   EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC,
+                   "Expected '~locales' filter to be an array, was: %s",
+                   node_str);
+      return FALSE;
+    }
+
+  locale_nodes = json_array_get_elements (skip_locales_array);
+
+  for (iter = locale_nodes; iter != NULL; iter = iter->next)
+    {
+      const gchar *string = json_node_get_string (iter->data);
+
+      if (string == NULL)
+        {
+          g_autofree gchar *node_str = json_node_to_string (filter_value);
+          g_set_error (error,
+                       EOS_UPDATER_ERROR,
+                       EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC,
+                       "Expected '~locales' array to contain string values, was: %s",
+                       node_str);
+          return FALSE;
+        }
+
+      if (g_strv_contains ((const gchar * const *) supported_languages, string))
+        {
+          *is_filtered = TRUE;
+          return TRUE;
+        }
+    }
+
+  *is_filtered = FALSE;
+  return TRUE;
+}
+
+static gboolean
+locale_only_filter_applies (JsonNode  *filter_value,
+                                  gboolean  *is_filtered,
+                                  GError   **error)
+{
+  JsonArray *skip_locales_array = json_node_get_array (filter_value);
+  g_autoptr(GList) locale_nodes = NULL;
+  GList *iter = NULL;
+  g_auto(GStrv) supported_languages = NULL;
+
+  g_return_val_if_fail (is_filtered != NULL, FALSE);
+
+  if (!get_locales_list_from_flatpak_installation (&supported_languages,
+                                                   error))
+    return FALSE;
+
+  if (skip_locales_array == NULL)
```

Comment:
Might as well do this `(skip_locales_array == NULL)` check before calling `get_locales_list_from_flatpak_installation()`, to avoid some wasted work on the error path.